### PR TITLE
feat(NOTIFY-1001): add granular account syncing

### PR DESCRIPTION
**Pull request by mathieuartu**
## Explanation

This PR adds complete support for the new [Account Syncing](https://www.figma.com/board/sx26FtZcQc0zs0z5hpdqBh/Account-Syncing-Solution?node-id=0-1&t=K7TdBL4Q5gO5hM78-0) feature.

Account syncing will happen only when a user unlocks their wallet for now. 
More hooks will be coming in the near future.

⚠️ **IMPORTANT NOTE**: Account syncing is disabled by default. This is enforced by the internal env variable `isAccountSyncingEnabled`. This env variable will be removed when account syncing is released.

## References

[NOTIFY-1001](https://consensyssoftware.atlassian.net/jira/software/projects/NOTIFY/boards/616?selectedIssue=NOTIFY-1001)

### `@metamask/profile-sync-controller`

- **ADDED**: added support for the new account syncing feature

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

[NOTIFY-1001]: https://consensyssoftware.atlassian.net/browse/NOTIFY-1001?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.test.ts [hunk 1]
```diff
@@ -444,20 +1290,40 @@ function mockUserStorageMessenger() {
...
+      nameLastUpdatedAt: 1,
+    },
+  });
+
   jest.spyOn(messenger, 'call').mockImplementation((...args) => {
     const [actionType, params] = args;
     if (actionType === 'SnapController:handleRequest') {
-      if (params?.request.method === 'getPublicKey') {
+      if (
+        (params as Exclude<typeof params, string | number>)?.request.method ===
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.test.ts:1310]: yep get it, the types for this mock are a little annoying. Will investigate.

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts [hunk 2]
```diff
@@ -56,11 +68,16 @@ export type UserStorageControllerState = {
    * Loading state for the profile syncing update
    */
   isProfileSyncingUpdateLoading: boolean;
+  /**
+   * Loading state for the account syncing update
+   */
+  isUserStorageAccountSyncingInProgress: boolean;
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts:74]: What do we need this loading state for in the UI.
(not against if need to attach this to the UI in future)

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts [hunk 3]
```diff
@@ -183,6 +232,90 @@ export default class UserStorageController extends BaseController<
...
+      isUserStorageAccountSyncingInProgress: boolean,
+    ) => {
+      // Publish event
+      const eventToPublish = isUserStorageAccountSyncingInProgress
+        ? 'UserStorageController:accountSyncingInProgress'
+        : 'UserStorageController:accountSyncingComplete';
+
+      this.messagingSystem.publish(
+        eventToPublish,
+        isUserStorageAccountSyncingInProgress,
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts:249]: Nit - idk if we need to publish this boolean.

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts [hunk 4]
```diff
@@ -183,6 +232,90 @@ export default class UserStorageController extends BaseController<
...
+        address,
+      );
+
+      if (!internalAccount) {
+        return;
+      }
+
+      // Map the internal account to the user storage account schema
+      const mappedAccount =
+        mapInternalAccountToUserStorageAccount(internalAccount);
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts:289]: This is nice.

Compared to network syncing, we can map the shape to a small shape with only the relevant info.

I don't foresee the accounts shape changing, so we don't need to handle versioning.

There will be a future item to handle versioning as a whole (alongside network syncing)

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts [hunk 5]
```diff
@@ -183,6 +232,90 @@ export default class UserStorageController extends BaseController<
...
+        mapInternalAccountToUserStorageAccount,
+      );
+
+      await Promise.all(
+        userStorageAccountsList.map(async (userStorageAccount) => {
+          await this.performSetStorage(
+            `accounts.${userStorageAccount.a}`,
+            JSON.stringify(userStorageAccount),
+          );
+        }),
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts:314]: Future - asked the BE team to provide a batch endpoint so we don't need to perform multiple API calls here.

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts [hunk 6]
```diff
@@ -205,9 +338,15 @@ export default class UserStorageController extends BaseController<
...
-      this.messagingSystem.subscribe('KeyringController:unlock', () => {
-        this.#isUnlocked = true;
-      });
+      this.messagingSystem.subscribe(
+        'KeyringController:unlock',
+        // eslint-disable-next-line @typescript-eslint/no-misused-promises
+        async () => {
+          this.#isUnlocked = true;
+
+          await this.syncInternalAccountsWithUserStorage();
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts:347]: Love this pattern!
GL if we can find an event for extension/mobile open.
I'm still not against a CRON (every 5mins/1hr for eventual consistency).

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts [hunk 7]
```diff
@@ -220,6 +359,9 @@ export default class UserStorageController extends BaseController<
   constructor(params: {
     messenger: UserStorageControllerMessenger;
     state?: UserStorageControllerState;
+    env?: {
+      isAccountSyncingEnabled?: boolean;
+    };
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts:364]: Perfect!

I prefer this as we can then allow platforms to control this (for worst case of easy rollback)

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts [hunk 8]
```diff
@@ -496,4 +652,167 @@ export default class UserStorageController extends BaseController<
       state.isProfileSyncingUpdateLoading = isProfileSyncingUpdateLoading;
     });
   }
+
+  /**
+   * Syncs the internal accounts list with the user storage accounts list.
+   * This method is used to make sure that the internal accounts list is up-to-date with the user storage accounts list and vice-versa.
+   * It will add new accounts to the internal accounts list, update/merge conflicting names and re-upload the results in some cases to the user storage.
+   */
+  async syncInternalAccountsWithUserStorage(): Promise<void> {
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts:661]: Nice func. Logic is consistent with what was described in PRD and tech doc.

Future - I feel like I want to pull out this logic to keep the controller thin. We can do this later (need to see how bloated this file gets once we add network syncing).

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts [hunk 9]
```diff
@@ -496,4 +652,167 @@ export default class UserStorageController extends BaseController<
...
+      // Get the internal accounts list again since new accounts might have been added in the previous step
+      internalAccountsList = await this.#accounts.getInternalAccountsList();
+
+      for await (const internalAccount of internalAccountsList) {
+        const userStorageAccount = userStorageAccountsList.find(
+          (account) => account.a === internalAccount.address,
+        );
+
+        if (!userStorageAccount) {
+          await this.#accounts.saveInternalAccountToUserStorage(
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts:717]: Future/perf - we can refactor this once we support batch.

We'll be mindful if there are users with many accounts.

File: packages/profile-sync-controller/src/controllers/user-storage/accounts/constants.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
+export const USER_STORAGE_VERSION = '1';
+
+// Force cast. We don't really care about the type here since we treat it as a unique symbol
+export const USER_STORAGE_VERSION_KEY: unique symbol = 'v' as never;
+
+export const LOCALIZED_DEFAULT_ACCOUNT_NAMES = [
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/accounts/constants.ts:6]: lol

Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/controllers/user-storage/accounts/constants.ts:6]: it is what it is.

Maybe we can add a comment on why this was needed :)

File: packages/profile-sync-controller/src/controllers/user-storage/accounts/user-storage.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  USER_STORAGE_VERSION,
+  LOCALIZED_DEFAULT_ACCOUNT_NAMES,
+} from './constants';
+
+export type UserStorageAccount = {
+  /**
+   * The Version 'v' of the User Storage.
+   * NOTE - will allow us to support upgrade/downgrades in the future
+   */
+  [USER_STORAGE_VERSION_KEY]: typeof USER_STORAGE_VERSION;
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/accounts/user-storage.ts:14]: Future - hmmm I think we may need to have different version for each different feature.

We can couple this work as part of the versioning support ticket.

**Review by Prithpal-Sooriya — DISMISSED** [review]
LGTM, some comments to resolve before merging.

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.test.ts (hunk 1 above)
Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.test.ts:1310]: Yep, `messenger.call` unions the actions and parameters which makes it hard grab the right type.

Use this to shim in the correct types.

```ts
  // Creates the correct typed call params for mocks
  type CallParams = {
    [K in AllowedActions['type']]: [
      K,
      ...Parameters<Extract<AllowedActions, { type: K }>['handler']>,
    ];
  }[AllowedActions['type']];
  
  ...
  const [actionType, Params] = args as unknown as CallParams;
```

Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.test.ts:1310]: ***Correction

The underlying `call` method is typed correctly, but I'm suspecting the `jest.spyOn` is not playing well with generic type parameters and is "expanding" the type, hence the CallParams are a union.

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts (hunk 2 above)
Comment (mathieuartu) [reply packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts:74]: @Prithpal-Sooriya I'm not even sure, I added it for the sake of maybe product asking for a loading indicator, but I didn't hear back on this, so maybe best to delete this (and maybe re-add later)? WDYT?

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts (hunk 8 above)
Comment (mathieuartu) [reply packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts:661]: Me too, we should definitely extract it like you guys did with set/get storage methods. Will do that in a later PR if that works for you @Prithpal-Sooriya ?

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.test.ts (hunk 1 above)
Comment (mathieuartu) [reply packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.test.ts:1310]: Nice one @Prithpal-Sooriya , added!

File: packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts (hunk 8 above)
Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts:661]: Yep later!

**Review by Prithpal-Sooriya — APPROVED** [review]

**Review by Prithpal-Sooriya — APPROVED** [review]

**Pull request merged**